File: manual/code/generate_strategies/IR(pseudocode)/python/python/function_pycrypto_decrypt.py
from Crypto.Cipher import AES
import base64
import os

class AESCipher:
    def process(key: str, text: str, mode: str = 'encrypt') -> str:
        key = key.ljust(32)[:32] 
        if mode == 'encrypt':
            block_size = AES.block_size
            padding_length = block_size - (len(text) % block_size)
            padded_text = text + (chr(padding_length) * padding_length)

            iv = os.urandom(AES.block_size)  
            cipher = AES.new(key.encode('utf-8'), AES.MODE_CBC, iv)
            ciphertext = cipher.encrypt(padded_text.encode('utf-8'))
            return base64.b64encode(iv + ciphertext).decode('utf-8')

        elif mode == 'decrypt':
            encrypted_text = base64.b64decode(text)
            iv = encrypted_text[:AES.block_size] 
            cipher = AES.new(key.encode('utf-8'), AES.MODE_CBC, iv)
            decrypted_text = cipher.decrypt(encrypted_text[AES.block_size:]).decode('utf-8')

            padding_length = ord(decrypted_text[-1])
            if padding_length > AES.block_size:
                raise ValueError("Invalid padding")
            unpadded_text = decrypted_text[:-padding_length]
            return unpadded_text

        else:
            raise ValueError("Mode must be 'encrypt' or 'decrypt'")

File: data/cpp/src/function_pycrypto_decrypt.cpp
#include <string>
#include <cryptopp/aes.h>
#include <cryptopp/modes.h>
#include <cryptopp/filters.h>
#include <cryptopp/base64.h>
#include <cryptopp/osrng.h>
#include <stdexcept>
class AESCipher {
public:
    static std::string process(const std::string& key, const std::string& text, const std::string& mode = "encrypt") {
        std::string paddedKey = key;
        paddedKey.resize(32, ' ');
        paddedKey = paddedKey.substr(0, 32);
        
        const int blockSize = CryptoPP::AES::BLOCKSIZE;
        
        if (mode == "encrypt") {
            int paddingLength = blockSize - (text.length() % blockSize);
            std::string paddedText = text;
            for (int i = 0; i < paddingLength; i++) {
                paddedText.push_back(static_cast<char>(paddingLength));
            }

            CryptoPP::AutoSeededRandomPool rng;
            CryptoPP::byte iv[blockSize];
            rng.GenerateBlock(iv, blockSize);
            
            CryptoPP::CBC_Mode<CryptoPP::AES>::Encryption encryptor;
            encryptor.SetKeyWithIV((CryptoPP::byte*)paddedKey.data(), paddedKey.size(), iv);
            
            std::string ciphertext;
            CryptoPP::StringSource(paddedText, true, 
                new CryptoPP::StreamTransformationFilter(encryptor,
                    new CryptoPP::StringSink(ciphertext)
                )
            );
            
            std::string ivAndCiphertext;
            ivAndCiphertext.append(reinterpret_cast<const char*>(iv), blockSize);
            ivAndCiphertext.append(ciphertext);
            
            std::string encoded;
            CryptoPP::StringSource(ivAndCiphertext, true,
                new CryptoPP::Base64Encoder(
                    new CryptoPP::StringSink(encoded),
                    false 
                )
            );
            
            return encoded;
        } 
        else if (mode == "decrypt") {
            try {
                std::string decoded;
                CryptoPP::StringSource(text, true,
                    new CryptoPP::Base64Decoder(
                        new CryptoPP::StringSink(decoded)
                    )
                );
                
                if (decoded.length() < blockSize) {
                    throw std::invalid_argument("Ciphertext too short");
                }
                
                std::string iv = decoded.substr(0, blockSize);
                std::string ciphertext = decoded.substr(blockSize);
                
                CryptoPP::CBC_Mode<CryptoPP::AES>::Decryption decryptor;
                decryptor.SetKeyWithIV((CryptoPP::byte*)paddedKey.data(), paddedKey.size(), 
                                      (CryptoPP::byte*)iv.data());
                
                std::string decrypted;
                CryptoPP::StringSource(ciphertext, true,
                    new CryptoPP::StreamTransformationFilter(decryptor,
                        new CryptoPP::StringSink(decrypted)
                    )
                );
                
                if (decrypted.empty()) {
                    throw std::invalid_argument("Empty decrypted text");
                }
                
                unsigned char paddingLength = static_cast<unsigned char>(decrypted[decrypted.length() - 1]);
                if (paddingLength > blockSize) {
                    throw std::invalid_argument("Invalid padding");
                }
                
                return decrypted.substr(0, decrypted.length() - paddingLength);
            } 
            catch (const CryptoPP::Exception& e) {
                throw std::invalid_argument(std::string("Decryption failed: ") + e.what());
            }
            catch (const std::exception& e) {
                throw std::invalid_argument(std::string("Decryption failed: ") + e.what());
            }
        } 
        else {
            throw std::invalid_argument("Mode must be 'encrypt' or 'decrypt'");
        }
    }
};